clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/AggKafkaConsumerService.java

```diff
@@ -100,6 +131,64 @@ public void stopInner() throws Exception {
     for (KafkaConsumerService consumerService: kafkaServerToConsumerServiceMap.values()) {
       consumerService.stop();
     }
+    if (this.stuckConsumerRepairExecutorService != null) {
+      this.stuckConsumerRepairExecutorService.shutdownNow();
+    }
+  }
+
+  protected static Runnable getStuckConsumerDetectionAndRepairRunnable(
+      Map<String, KafkaConsumerService> kafkaServerToConsumerServiceMap,
+      Map<String, StoreIngestionTask> versionTopicStoreIngestionTaskMapping,
+      long stuckConsumerRepairThresholdMs,
+      StuckConsumerRepairStats stuckConsumerRepairStats) {
+    return () -> {
+      /**
+       * The following logic can be further optimized in the following way:
+       * 1. If the max delay of previous run is much smaller than the threshold.
+       * 2. In the next run, the max possible delay will be schedule interval + previous max delay ms, and if it is
+       * still below the threshold, this function can return directly.
+       * We are not adopting such optimization right now because:
+       * 1. Extra state to maintain.
+       * 2. The schedule interval is supposed to be high.
+       * 3. The check is cheap when there is no stuck consumer.
+       */
+      boolean scanStoreIngestionTaskToFixStuckConsumer = false;
+      for (KafkaConsumerService consumerService: kafkaServerToConsumerServiceMap.values()) {
+        long maxDelayMs = consumerService.getMaxElapsedTimeMSSinceLastPollInConsumerPool();
+        if (maxDelayMs >= stuckConsumerRepairThresholdMs) {
+          scanStoreIngestionTaskToFixStuckConsumer = true;
+          LOGGER.warn("Found some consumer has stuck for {} ms, will start the repairing procedure", maxDelayMs);
+          break;
+        }
+      }
+      if (!scanStoreIngestionTaskToFixStuckConsumer) {
+        return;
+      }
+      stuckConsumerRepairStats.recordStuckConsumerFound();
+
+      AtomicBoolean repairSomeIngestionTask = new AtomicBoolean(false);
+      versionTopicStoreIngestionTaskMapping.forEach((vt, sit) -> {
+        if (!sit.isProducingTopicHealthy()) {
+          LOGGER.info(
+              "The ingestion topics (version topic and/or real-time topic) are not healthy for "
+                  + "store version: {}, will kill the ingestion task to try to unblock shared consumer",
+              vt);
+          /**
+           * The following function call will interrupt all the stuck {@link org.apache.kafka.clients.producer.KafkaProducer#send} call
+           * to non-existing topics.
+           */
+          sit.closeVeniceWriters(false);
+          sit.kill();
```

Comment:
I checked `KafkaStoreIngestionService.killConsumptionTask` and there seems to have some clean up work after calling `StoreIngestionTask.kill`. Are those cleanup steps necessary for here as well?

Reply:
Does `versionTopicStoreIngestionTaskMapping` contain all possible SIT on this server? Just to think out loud, do we need to cross-check that the SIT that's going to be killed does indeed have a topic partition assigned to the slow `KafkaConsumer` (or `ConsumptionTask`), which causes `getMaxElapsedTimeMSSinceLastPollInConsumerPool` to alarm in the first place. My worry is that if isProducingTopicHealthy? or in general, it is okay to kill any SIT as long as it doesn't pass the  `isProducingTopicHealthy` test

Reply:
This is a good question.
`KafkaStoreIngestionService.killConsumptionTask` does more cleanup than the `kill`, such as cleanup locks and handle to the ingestion task, and I should potentially call `KafkaStoreIngestionService#killConsumptionTask` instead.

`versionTopicStoreIngestionTaskMapping` contains all the live `SIT` on the same server.

Regarding the following:
```
Just to think out loud, do we need to cross-check that the SIT that's going to be killed does indeed have a topic partition assigned to the slow KafkaConsumer (or ConsumptionTask), which causes getMaxElapsedTimeMSSinceLastPollInConsumerPool to alarm in the first place. My worry is that if isProducingTopicHealthy? or in general, it is okay to kill any SIT as long as it doesn't pass the isProducingTopicHealthy test
```
I thought about it during the implementation, and checking the attached SITs for a particular stuck consumer is more complex because with partition-wise shared consumers, one SIT can be assigned to multiple consumers, and in consumer-level, there is no concept of `SIT`, but `ConsumedDataReceiver`.

In the high-level, I think the ingestion task should be killed if the producing topics are deleted, such as RT or VT, and I guess you might be worrying that this can potentially trigger some edge cases, such as `TopicManager#containsTopic` return `false`, but the topic does exist somehow.

I am thinking adding some retries into the function: `isProducingTopicHealthy` to make it more solid.